src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,358 @@
+		pdata->ticks_to_trigger = 0;
+		task->state = SOF_TASK_STATE_RUNNING;
+		k_sem_give(&pdata->sem);
+	} else {
+		/* wait for tick */
+		pdata->ticks_to_trigger	= start / LL_TIMER_PERIOD_US;
+		task->state = SOF_TASK_STATE_QUEUED;
+	}
+
+	scheduler_dp_unlock(lock_key);
+
+	return 0;
+}
+
+static struct scheduler_ops schedule_dp_ops = {
+	.schedule_task		= scheduler_dp_task_shedule,
+	.schedule_task_cancel	= scheduler_dp_task_cancel,
+	.schedule_task_free	= scheduler_dp_task_free,
+};
+
+int scheduler_dp_init_secondary_core(void)
+{
+	if (!dp_sch)
+		return -ENOMEM;
+
+	/* register the scheduler instance for secondary core */
+	scheduler_init(SOF_SCHEDULE_DP, &schedule_dp_ops, dp_sch);
+
+	return 0;
+}
+
+int scheduler_dp_init(void)
+{
+	dp_sch = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*dp_sch));
+	if (!dp_sch)
+		return -ENOMEM;
+
+	list_init(&dp_sch->tasks);
+
+	scheduler_init(SOF_SCHEDULE_DP, &schedule_dp_ops, dp_sch);
+
+	notifier_register(NULL, NULL, NOTIFIER_ID_LL_POST_RUN, scheduler_dp_ll_tick, 0);
+
+	return 0;
+}
+
+int scheduler_dp_task_init(struct task *task,
+			   const struct sof_uuid_entry *uid,
+			   const struct task_ops *ops,
+			   void *data,
+			   uint16_t core,
+			   size_t stack_size,
+			   uint32_t task_priority)
+{
+	struct task_dp_pdata *pdata = NULL;
+	struct k_thread *p_thread = NULL;
+	k_tid_t thread_id = NULL;
+	void *p_stack = NULL;
+	int ret;
+
+	if (task->priv_data)
+		return -EEXIST;
+
+	ret = schedule_task_init(task, uid, SOF_SCHEDULE_DP, 0, ops->run, data, core, 0);
+	if (ret < 0) {
+		tr_err(&dp_tr, "zephyr_dp_task_init(): schedule_task_init failed");
+		return ret;
+	}
+	task->ops.complete = ops->complete;
+	task->ops.get_deadline = ops->get_deadline;
+
+	/* allocate priv data space */
+	pdata = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM,
+			sizeof(*pdata));
+	if (!pdata) {
+		tr_err(&dp_tr, "zephyr_dp_task_init(): pdata alloc failed");
+		ret = -ENOMEM;
+		goto err;
+	}
+	k_sem_init(&pdata->sem, 0, 1);
+
+	/* allocate stack for a thread */
+	stack_size = Z_KERNEL_STACK_SIZE_ADJUST(stack_size);
+	p_stack = rballoc_align(0, SOF_MEM_CAPS_RAM, stack_size, Z_KERNEL_STACK_OBJ_ALIGN);
+	if (!p_stack) {
+		tr_err(&dp_tr, "zephyr_dp_task_init(): stack alloc failed");
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/* allocate zephyr thread  */
+	p_thread = rzalloc(SOF_MEM_ZONE_SYS_SHARED, 0, SOF_MEM_CAPS_RAM,
+			   sizeof(*p_thread));
+	if (!p_thread) {
+		tr_err(&dp_tr, "zephyr_dp_task_init(): thread alloc failed");
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/* create a zephyr thread for the task */
+	thread_id = k_thread_create(p_thread, p_stack, stack_size, dp_thread_fn, task,
+				    NULL, NULL, task_priority, K_USER, K_FOREVER);
+	if (!thread_id)	{
+		ret = -EFAULT;
+		tr_err(&dp_tr, "zephyr_dp_task_init(): zephyr thread create failed");
+		goto err;
+	}
+	/* pin the thread to specific core */
+	ret = k_thread_cpu_pin(thread_id, task->core);
+	if (ret < 0) {
```

Comment:
Ditto, this won't fail unless something specifies a bad CPU or tries to pin a running thread.

Reply:
anyway, worth to check